### ISSUE TITLE
When called consecutively, HasFeatureFlag method gets an error on the second run

### ISSUE THREAD

**Issue by sevikarakose** [BulletTrainHQ/bullet-train-dotnet-client#5]
I've been trying 2.0.0 BulletTrain Client SDK on a .NetCore 3.1 WebApi project and on a Console application. For Api calls, my initial plan was to run an on-prem version on Docker. But also I tried "https://app.bullet-train.io/" with a free account. 

- Console Application works properly. But WebApi project fails when a second call is made to the BulletTrain Api.

- So here's my sample code for Console App, 

```
            BulletTrainConfiguration configuration = new BulletTrainConfiguration()
            {
                ApiUrl = "http://localhost:8000/api/v1/",
                EnvironmentKey = "my-local-env-key"
            };

            BulletTrainClient bulletClient = new BulletTrainClient(configuration);
            bool featureEnabled = await bulletClient.HasFeatureFlag("is_it_weekend");
                if (featureEnabled) {
                    // Run the code to execute enabled feature
                    Console.WriteLine("Hello World!");
                } else {
                    // Run the code if feature switched off
                    Console.WriteLine("GoodBye World!");
                }
```

- Here is the sample code from the WebApi project, they're almost identical (I used BulletTrain docs for implementation)

```
            BulletTrainConfiguration configuration = new BulletTrainConfiguration()
            {
                ApiUrl = "http://localhost:8000/api/v1/",
                EnvironmentKey = "my-local-env-key"
            };

            BulletTrainClient bulletClient = new BulletTrainClient(configuration);
            bool featureEnabled = await bulletClient.HasFeatureFlag("is_it_weekend");
            if (featureEnabled) {
                // Run the code to execute enabled feature
                return Ok(myWeekendItem);
            } else {
                // Run the code if feature switched off
                return NotFound();
            }
```

- When I run it for the first time, it works properly. Without even changing anything, when I run on the web api the second time, it throws the below exception:

Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: Error: An unhandled exception has occurred while executing the request.

System.NullReferenceException: Object reference not set to an instance of an object.
   at BulletTrain.BulletTrainClient.GetFeatureFlags(String identity)
   at BulletTrain.BulletTrainClient.HasFeatureFlag(String featureId, String identity)
   at Glossary.Controllers.GlossaryController.Get(String term) in /Users/sevikarakose/Projects/Glossary/Controllers/GlossaryController.cs:line 52
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

- This also happens when I try it with the "https://app.bullet-train.io/" for api URL's. It looks like it's about dotnet-client.

**Comment by dabeeeenster** [comment]
Hi @sevikarakose - we'll take a look at this.

**Comment by ssg-luke** [comment]
@sevikarakose Hi, can I just double check, you are only creating configuration and initialising an instance of BulletTrainClient **once** in your Web API right? When you say run on the web api the 2nd time, you mean you fully terminate your api and re-start it?

**Comment by sevikarakose** [comment]
Hi there, thanks for the quick reply. I do not terminate my api and restart it. I call my method that goes to the feature and check it while the app is still running. 

Actually I was doing a POC with the dotnet client for my company, so I created a sample api and saw it worked with (let's say) feature flag on. Then without restarting it, I went to the BulletTrain front end (running in Docker) changed the flag value and but this time, I got the error that I pasted.

So to double check, I restarted my api, got the flag result from the BulletTrain api the first time again successfully (whatever the result is), but then I did not change the flag's value from BulletTrain front end, yet again the api failed.

So to triple check :), I created a Console App, and after seeing it's working properly, I put my method in a loop for 1000 times to go check the value from the BulletTrain api. While it was working successfully, I went to the BulletTrain frontend and changed the value for a few times. It never failed, always got the proper results.

**Comment by ssg-luke** [comment]
Ok thanks @sevikarakose thats really useful. Seems as though any testing with a Console App is not going to be useful in this case. I've not been able to replicate the error you're seeing in the Console app. Would it be possible for you to send across the PoC Web API you've been working on?

**Comment by sevikarakose** [comment]
Hi there! Thanks again for your fast response. You can clone this repository to check it out!
https://github.com/sevikarakose/Glossary
Before running it, you will need to change the env key for BulletTrain api. The feature I check is called, "show_feature"
When it's up and running, just go to one of these url's for two or more times to see the problem
http://localhost:5000/api/glossary
https://localhost:5001/api/glossary

**Comment by ssg-luke** [comment]
Thanks for that! I replicated the issue straight away. The reason you were seeing an exception is that BulletTrainClient is only supposed to be created and initialised once for the application. I have released v2.0.1 which enforces this and it will now throw a NotSupportedException error. I have also added for static variable `instance` for much greater convenience. After creating a `new BulletTrainClient()` during startup of the app you can now use it anywhere by using `BulletTrainClient.instance`

**Comment by sevikarakose** [comment]
Hey, that's great news! 🤩 I suggested this might be the problem, but searched in the documentations thoroughly (Nuget.org & BulletTrain front-end code samples & github documentations) and when I didn't find anything about it, I just went with it. Do you plan to change the documentations as well? Cheers and thank you!
